src/systemd.c

```diff
@@ -51,7 +51,19 @@ static int systemd_fd = -1;
 static const char *sd_unix_path = NULL;
 
 int systemd_init(void) {
-	sd_unix_path = getenv("NOTIFY_SOCKET");
+	bool exceeded;
+	/* To be sure, that path to unix socket is OK */
+	struct sockaddr_un sa = {
+		.sun_family = AF_UNIX,
+		.sun_path = { '\0' }
+	};
+	sd_unix_path = getenv_safe("NOTIFY_SOCKET", sizeof(sa.sun_path),
+				   &exceeded);
+	if (exceeded) {
+		say_error("systemd: NOTIFY_SOCKET is longer than "
+			  "MAX_UNIX_PATH");
+		goto error;
+	}
```

Comment:
Not really.
We checked that `sd_unix_path != NULL`, so now it's a string of length at least zero (meaning `sd_unix_path[0] == '\0'`).
Now in case we have a string of length zero (the only case when sd_unix_path[1] is invalid),
`sd_unix_path[0] != '@' && sd_unix_path[0] != '/'` evaluate to true and the part with `|| sd_unix_path[1] == '\0'` is not executed. So we should be safe there.

Reply:
I see now, thanks for the clarification!